### ISSUE TITLE
DOMDocument is required... RedHat: that is a problem

### ISSUE THREAD

**Issue by loushou** [woothemes/woocommerce#7404]
So, our dev server had a whoops recently (unrelated anything here). Consequently, I had to provision a new dev server, and get it up and running with all our dev sites again. This was a fresh RedHat installation. We have a plugin that is built on top of WooCommerce, called OpenTickets. With the new release of WC2.3, we have started a QA round to track down any new unknown issues in our software. As a result, we found a bug, which is not an OpenTickets bug, but rather what I believe to be a minor WooCommerce oversight, leading to a fatal error on some RedHat Enterprise Linux system. Side-note: I am honestly not sure how long this has been around, but I can say with 100% certainty that it does manifest with newly installed RHEL server that does not specifically install the php-dom package.

I have already solved this on my server; however, RHEL and Centos implementations of php5.4+ *do not* come with the DOMDocument php extension by default. You have to install a separate package for this, php-dom. Consequently, before solving this on my server, I was receiving this error:

`Fatal error: Class 'DOMDocument' not found in /path/to/webroot/wp-content/plugins/woocommerce/includes/libraries/class-emogrifier.php on line 515`

Obviously, any release producing any error, especially on a widely used production ready version of linux like RedHat Linux Enterprise, is not preferable. Now, from the look of it, woocommerce/includes/libraries/class-emogrifier.php is an external library, making it less likely to be modified to solve such a problem; however, given the fact that a dependency of this library is not included by default on RHEL... do you think that there should be at least some fallback, as to not produce a fatal error? 

The quickest, dirtiest solution would be to simply remove the requirement of DOMDocument from this class, by simply replacing it with SimpleXML, which has been around since early PHP5, well before the WP min php version. Looking at the code, this could be pretty straightforward, since none of the advanced DOMDocument features are actually being used here, and each used feature is already existent in simplexml. The downside is that this requires editing an external library.

A longer term, let obtrusive solution, would be to prefer this library, but have a duplicate library that uses simplexml instead. Then check on load if DOMDocument is installed, and use the preferred lib if it is, or use the safe fallback if not. This is admittedly a lot more work, but is probably preferable, unless you have direct access to change this library's code, without adverse ramifications. 

In short, I think the heart of this project is to reach as many people as possible. By requiring a php extension that is not guaranteed to be installed (since something like RHEL does not have it on by default), I think you are going against the spirit of the wide access approach. Granted, I was able to remedy the situation on my server by simply installing a package... but not everyone will have that luxury. The fix will take some work, but I think you will find that you will be less prone to server implementation problems with that effort put forward, and you will be able to reach more users with less problems.

Loushou

**Comment by claudiosmweb** [comment]
http://php.net/manual/en/dom.installation.php

**Comment by loushou** [comment]
Are you joking... I am aware that you can simply install the php extension. In fact, if you actually read what I wrote, I say that it is easily solved by doing so. But, apparently you also did not read that _not everyone has the ability to add a php extension_ ... blatantly ignoring a problem does not make it go away. Thanks for completely ignoring it. Great display of understanding your full target audience.

**Comment by mikejolley** [comment]
@loushou not everyone is using the build of RedHat you are using.

The emogrifier is the best option for our needs, are there are not really any alternatives to DOMDocument. We could add regex or something custom and funky but thats not really worth the development effort required.

Also from PHP.net:

> This extension is enabled by default. It may be disabled by using the following option at compile time: --disable-dom

So if your build disables a default PHP extension thats not really something we should be working around.

With the commit Claudio made, if you don't have DOMDocument, emails fall back to plain text. No breakage.

**Comment by claudiosmweb** [comment]
@loushou Yes I read, please see this:

SimpleXML: http://php.net/manual/en/simplexml.installation.php

DOM: http://php.net/manual/en/dom.installation.php

Both are enabled by default. change to SimpleXMLElement can lead to the same problem, because some server can be this disabled in your works: "php extension that is not guaranteed to be installed".

Now I'm warning you need it installed and was avoided faltal errors.

**Comment by loushou** [comment]
If you look at the code, the DOMDocument features that the library uses are not the advanced DOMDocument features. All those features are actually part of SimpleXML already. Also @mikejolley unless that is a bleed edge commit, it does not solve the problem, since this literally happened yesterday afternoon, with the latest version, minutes before I posted this bug. There was no fallback. It just fatal errored. Additionally, it seems like my whole post is not being read, so I will reiterate the whole problem:

_It is not just *my* build. This is the *standard* RedHat build. There is not any customizing going on here. It simply does not get installed with the RedHat implementation of php5.4 (or higher). You have to manually install it, which not everyone will have the ability to do._

@claudiosmweb then check for both, and choose which one is available, and fallback to a safe, non xml reliant version if neither is.

@both-of-you... at the very least, it should not throw a fatal error if DOMDocument is not installed; and just so we are clear, the current release DOES throw a fatal error, which is the whole point of this ticket. If we cannot agree on that, I'm not sure why I even wasted the time to post it here, or feed the troll regurgitating my own instructions of installing the DOMDocument php extension.

**Comment by loushou** [comment]
Actually. Just forget it. Obviously it is not a big enough issue to warrant attention, so I will just add information in our software's FAQs about how your software requires something they may not have installed. I concede.

**Comment by mikejolley** [comment]
@loushou You are aware Claudio closed this issue **with a commit which prevents the fatal error**. Maybe you missed that?... https://github.com/woothemes/woocommerce/commit/eff087f5ab780c197fd3e3d8283e7983cf1fd1a0

**Comment by claudiosmweb** [comment]
@mikejolley I don't think he saw my commit, probably he just wanting troll me.

@loushou Feel free to send a Pull Request with this DOMDocument/SimpleXML that you want.
